### PR TITLE
Add launch flow domain upsell

### PR DESCRIPTION
**Pull request by dzver**
#### Changes proposed in this Pull Request

This change adds a new step to the launch-site flow with a 50/50 split test for English locales.

#### Testing instructions

0. Make sure you are using an English locale
1. Sign up for a free user and site
2. Launch the site from the /home menu
3. Select a free domain and plan
4. For users in the test group, the new step should be visible
5. Select a domain. The active button should launch the site and direct you to checkout
6. There's no need to create a new site to test the skip button but it might be a good idea anyway. Remove the cart items, go to http://calypso.localhost:3000/start/launch-site/domains-launch?siteSlug=YOURSITE.wordpress.com and on step 4 - skip
7. Assign yourself to the control group and repeat. No additional steps should appear

Here is how it looks right now:

Screenshot 2020-12-03 at 16 36 10

Note: it's currently being discussed whether this particular experiment could be run with ExPlat. There is currently no easy way to do it so it's likely going to use the old system but ExPlat folks have been pinged.

### CODE REVIEW

File: client/signup/steps/free-plans-domain-discount-launch/index.jsx [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+				<div>
+					{ translate( 'Registration fee: {{strong}}%(originalCost)s{{/strong}}', {
+						args: {
+							originalCost: formatCurrency( suggestion.cost, currency, { stripZeros: true } ),
+						},
+						components: { strike: <strike />, strong: <strong /> },
+					} ) }
+				</div>
+				<small>
+					{ translate( 'Renews at: %(cost)s /year', {
```
Comment (a8ci18n) [inline client/signup/steps/free-plans-domain-discount-launch/index.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **22** times:
`translate( 'Renews at %(cost)s / year' )` *ES Score: 9*

File: client/signup/steps/free-plans-domain-discount-launch/index.jsx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+				<div>
+					{ translate( 'Registration fee: {{strong}}%(originalCost)s{{/strong}}', {
+						args: {
+							originalCost: formatCurrency( suggestion.cost, currency, { stripZeros: true } ),
+						},
+						components: { strike: <strike />, strong: <strong /> },
+					} ) }
+				</div>
+				<small>
+					{ translate( 'Renews at: %(cost)s / year', {
```
Comment (a8ci18n) [inline client/signup/steps/free-plans-domain-discount-launch/index.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **22** times:
`translate( 'Renews at %(cost)s / year' )` *ES Score: 9*

**Comment by davemart-in** [comment]
@dzver I reviewed the design and there isn't actually anything I'd change about the design for the v1 test. Nice work! 🙌

File: client/lib/signup/step-actions/index.js [hunk 3]
```diff
@@ -779,6 +804,27 @@ export function isPlanFulfilled( stepName, defaultDependencies, nextProps ) {
 	}
 }
 
+export function isFreePlansDomainUpselFulfilled( stepName, defaultDependencies, nextProps ) {
```
Comment (aneeshd16) [inline client/lib/signup/step-actions/index.js]: Small typo, should be `isFreePlansDomainUpsellFulfilled` 🙂

**Review by aneeshd16 — COMMENTED** [review]
I observed this possibly unintended behaviour:
1. Choose a domain from the last step of the launch flow.
2. On the checkout page, the domain should be visible in the cart. Press the browser back button.
3. Go through the launch flow and choose a different domain in the last step.
4. Both domains are now present in the cart on the checkout page.

File: client/signup/steps/free-plans-domain-discount-launch/index.jsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+									label={ FormattedSuggestion( translate, suggestion, index === 0 ) }
+									value={ suggestion.domain }
+								/>
+							</FormLabel>
+						) ) }
+					</FormFieldset>
+				) }
+				<div className="free-plans-domain-discount-launch__buttons">
+					<Button busy={ isLoading } primary onClick={ handleUpgradeButtonClick }>
+						{ isLoading ? '' : translate( 'Use %s', { args: [ selectedDomain ] } ) }
```
Comment (aneeshd16) [inline client/signup/steps/free-plans-domain-discount-launch/index.jsx]:
```suggestion
						{ isLoading ? translate( 'Loading' ) : translate( 'Use %s', { args: [ selectedDomain ] } ) }
```
Would it make sense to add the text "Loading" here?
During the loading state, it currently looks like this:
Kapture 2020-12-08 at 12 05 22

**Comment by aneeshd16** [comment]
I have updated the PR with changes and it is ready for a final round of reviews and testing.

File: client/signup/steps/free-plans-domain-discount-launch/index.jsx (hunk 4 above)
Comment (davemart-in) [reply client/signup/steps/free-plans-domain-discount-launch/index.jsx]: Yep. Adding "Loading" makes sense. 👍

File: client/signup/config/steps-pure.js [hunk 5]
```diff
@@ -114,6 +116,14 @@ export function generateSteps( {
 			dependencies: [ 'siteSlug' ],
 		},
 
+		'free-plans-domain-discount-launch': {
```
Comment (ddc22) [inline client/signup/config/steps-pure.js]: This name is a little confusing, maybe something like `domain-upsell` can be more descriptive and easier to understand.

File: client/signup/steps/free-plans-domain-discount-launch/index.jsx [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+		submitSignupStep( step, { selectedDomainUpsellItem: null } );
+		goToNextStep();
+	};
+
+	return (
+		<div className="free-plans-domain-discount-launch">
+			{ ! productData && <QuerySecureYourBrand domain={ domain } /> }
+			<Card style={ { maxWidth: '615px' } } className="free-plans-domain-discount-launch__card">
+				{ isLoading ? (
+					[ ...Array( 3 ) ].map( ( i ) => (
```
Comment (ddc22) [inline client/signup/steps/free-plans-domain-discount-launch/index.jsx]: Nice approach here! Also similar way to do this just for interest.
```javascript
Array( 3 ).fill().map( ( i ) => (
```

File: client/signup/config/flows.js [hunk 7]
```diff
@@ -39,7 +39,12 @@ function getCheckoutUrl( dependencies, localeSlug, flowName ) {
 }
 
 function dependenciesContainCartItem( dependencies ) {
-	return dependencies.cartItem || dependencies.domainItem || dependencies.themeItem;
+	return (
+		dependencies.cartItem ||
```
Comment (ddc22) [inline client/signup/config/flows.js:43]: Destructuring assignments on the function parameter itself can be used to reduce repetition here.

**Review by aneeshd16 — APPROVED** [review]

**Review by ddc22 — CHANGES_REQUESTED** [review]
Feature works as described and I believe will be a valuable feature to have. Some small feedback on the UI design and content.

image

Also added some code comments below.

File: client/signup/config/steps-pure.js (hunk 5 above)
Comment (aneeshd16) [reply client/signup/config/steps-pure.js]: That's a good suggestion! I am renaming it to `launch-flow-domain-upsell` for more clarity.

Comment (aneeshd16) [reply client/signup/config/steps-pure.js]: Updated the PR with this change.

Comment (ddc22) [reply client/signup/config/steps-pure.js]: Hmm, What if someone decides to use the this step in a different flow? I think steps should be agnostic of which flows they are in. wdyt?

Comment (aneeshd16) [reply client/signup/config/steps-pure.js]: > Hmm, What if someone decides to use the this step in a different flow? I think steps should be agnostic of which flows they are in. wdyt?

Agree with you on this. Good catch!

File: client/signup/steps/launch-flow-domain-upsell/index.jsx [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+				<div>
+					{ translate( 'Registration fee: {{strong}}%(originalCost)s{{/strong}}', {
+						args: {
+							originalCost: formatCurrency( suggestion.cost, currency, { stripZeros: true } ),
+						},
+						components: { strike: <strike />, strong: <strong /> },
+					} ) }
+				</div>
+				<small>
+					{ translate( 'Renews at: %(cost)s / year', {
```
Comment (a8ci18n) [inline client/signup/steps/launch-flow-domain-upsell/index.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **22** times:
`translate( 'Renews at %(cost)s / year' )` *ES Score: 9*

File: client/signup/steps/domain-upsell/index.jsx [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+				<div>
+					{ translate( 'Registration fee: {{strong}}%(originalCost)s{{/strong}}', {
+						args: {
+							originalCost: formatCurrency( suggestion.cost, currency, { stripZeros: true } ),
+						},
+						components: { strike: <strike />, strong: <strong /> },
+					} ) }
+				</div>
+				<small>
+					{ translate( 'Renews at: %(cost)s / year', {
```
Comment (a8ci18n) [inline client/signup/steps/domain-upsell/index.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **22** times:
`translate( 'Renews at %(cost)s / year' )` *ES Score: 9*

**Comment by aneeshd16** [comment]
> Feature works as described and I believe will be a valuable feature to have. Some small feedback on the UI design and content.
> 
> image
> 
> Also added some code comments below.

@davemart-in would be able to respond to this better!

**Review by ddc22 — APPROVED** [review]
Confidence checked feature after code changes
LGTM 🚢

**Pull request merged**